Mathlib/Tactic/Bound.lean

```diff
@@ -0,0 +1,280 @@
+
+/-- Branch on `1 ≤ a ∨ a ≤ 1` for `a^n` -/
+lemma pow_le_pow_right_of_le_one_or_one_le {R : Type} [OrderedSemiring R] {a : R} {n m : ℕ}
+    (h : 1 ≤ a ∧ n ≤ m ∨ 0 ≤ a ∧ a ≤ 1 ∧ m ≤ n) : a ^ n ≤ a ^ m := by
+  rcases h with ⟨a1, nm⟩ | ⟨a0, a1, mn⟩
+  · exact pow_le_pow_right a1 nm
+  · exact pow_le_pow_of_le_one a0 a1 mn
+
+-- Register guessing rules
+attribute [bound]
+  -- Which side of the `max` should we use as the lower bound?
+  le_max_of_le_left_or_le_right
+  lt_max_of_lt_left_or_lt_right
+  -- Which side of the `min` should we use as the upper bound?
+  min_le_of_left_le_or_right_le
+  min_lt_of_left_lt_or_right_lt
+  -- Given `a^m ≤ a^n`, is `1 ≤ a` or `a ≤ 1`?
+  pow_le_pow_right_of_le_one_or_one_le
+
+end Guessing
+
+/-!
+### Closing tactics
+-/
+
+/-- Close numerical goals with `norm_num` -/
+def boundNormNum : Aesop.RuleTac :=
+  Aesop.SingleRuleTac.toRuleTac fun i => do
+    let tac := do Mathlib.Meta.NormNum.elabNormNum .missing .missing .missing
+    let goals ← Lean.Elab.Tactic.run i.goal tac |>.run'
+    if !goals.isEmpty then failure
+    return (#[], none, some .hundred)
+attribute [aesop unsafe 10% tactic (rule_sets := [Bound])] boundNormNum
+
+/-- Close numerical and other goals with `linarith` -/
+def boundLinarith : Aesop.RuleTac :=
+  Aesop.SingleRuleTac.toRuleTac fun i => do
+    Linarith.linarith false [] {} i.goal
+    return (#[], none, some .hundred)
+attribute [aesop unsafe 5% tactic (rule_sets := [Bound])] boundLinarith
+
+/-!
+### `bound` tactic implementation
+-/
+
+/-- Pull the array out of an optional `"[" term,* "]"` syntax, or return `#[]` -/
+def maybeTerms : Syntax → Array Syntax
+  | node _ _ #[_, node _ _ s, _] => s.getEvenElems
+  | _ => #[]
+
+/-- Add each provided hypothesis `x` to the context, as by with `have := x` -/
+def addHyps (xs : Array Syntax) : TacticM Unit :=
+  if xs.isEmpty then pure () else Tactic.withMainContext do
+    for x in xs do
+      let v ← elabTerm x none
+      let t ← inferType v
+      liftMetaTactic fun g ↦ do
+        let g ← g.assert `h t v
+        let (_, g) ← g.intro1P
+        return [g]
+
+/-- Aesop configuration for `bound` -/
+def boundConfig : Aesop.Options := {
+  enableSimp := false
+}
+
+end Bound
+
+/-- `bound` tactic for proving inequalities via straightforward recursion on expression structure.
+
+`bound` is built on top of `aesop`, and uses
+1. Apply lemmas registered via the `@bound` attribute
+2. Forward lemmas registered via the `@bound_forward` attribute
+3. Local hypotheses from the context
+4. Optionally: additional hypotheses provided as `bound [h₀, h₁]` or similar. These are added to the
+   context as if by `have := hᵢ`. -/
+elab "bound" lemmas:(("[" term,* "]")?) : tactic => do
+  Bound.addHyps (Bound.maybeTerms lemmas)
+  let tac ← `(tactic| aesop (rule_sets := [Bound, -default]) (config := Bound.boundConfig))
+  liftMetaTactic fun g ↦ do return (← Lean.Elab.runTactic g tac.raw).1
```

Comment:
```suggestion
syntax "bound " (" [" term,* "]")? : tactic

elab_rules : tactic
  | `(tactic| bound $[[$lemmas:term,*]]?) => do
    let lemmas := (lemmas.getD ⟨#[]⟩).getElems
    Bound.addHyps lemmas
    let tac ← `(tactic| aesop (rule_sets := [Bound, -default]) (config := Bound.boundConfig))
    liftMetaTactic fun g ↦ do return (← Lean.Elab.runTactic g tac.raw).1
```
This should allow you to avoid `maybeTerms`: this leverages `.getElems`, which "does the right thing".